src/msg/xio/XioMessenger.cc

```diff
@@ -403,6 +409,32 @@ int XioMessenger::pool_hint(uint32_t dsize) {
 				   XMSG_MEMPOOL_QUANTUM, 0);
 }
 
+int XioMessenger::get_nconns_per_portal(const string &mname)
+{
+  int nconns_per_portal;
+
+  if (mname == "client" || mname == "cluster" || mname == "mon")
+    nconns_per_portal = max(cct->_conf->xio_max_conns_per_portal, 100);
+  else if (mname == "hb_back_server" || mname == "hb_front_server" ||
+           mname == "hbclient")
+    // osd_heartbeat_min_peers * 2 (hb_back & hb_front) * 2 (leading conns)
+    nconns_per_portal = cct->_conf->osd_heartbeat_min_peers * 4;
+  else /* osd's ms_objecter or client messenger */
+    nconns_per_portal = min(cct->_conf->xio_max_conns_per_portal, 32);
+
+  return nconns_per_portal;
+}
+
+int XioMessenger::get_nportals(const string &mname)
+{
+  int nportals = 1;
+
+  if (mname == "client" || mname == "cluster")
+    nportals = max(cct->_conf->xio_portal_threads, 1);
+
+  return nportals;
+}
+
```

Comment:
@cbodley, @vuhuong, @liewegas, @gregsfortytwo Agree with Casey this feels like policy, and more importantly, it doesn't seem like we want the messenger "names" in the interface.  Is there a way we can make some XioMessenger-specific policy that does the work explicitly, so we avoid names and also avoid changing shared policy structures (which we pretty much can't change 'till Kraken)?